File: src/Utilities/BinaryTree.cpp
#include "BinaryTree.h"

#include <iostream>
#include <tuple>

using namespace std;

TreeNode* createBinaryTreeNode( int value ) {
    TreeNode* pNode = new TreeNode( value );
    return pNode;
}

void connectTreeNodes( TreeNode* pParent, TreeNode* pLeft, TreeNode* pRight ) {
    if ( pParent != nullptr ) {
        pParent->left = pLeft;
        pParent->right = pRight;
    }
}

void printTreeNode( const TreeNode* pNode ) {
    if ( pNode != nullptr ) {
        cout << "value of this node is: " << pNode->val << endl;

        if ( pNode->left != nullptr )
            cout << "value of its left child is: " << pNode->left->val << endl;
        else
            cout << "left child is nullptr.\n";

        if ( pNode->right != nullptr )
            cout << "value of its right child is: " << pNode->right->val
                 << endl;
        else
            cout << "right child is nullptr.\n";
    } else {
        cout << "this node is nullptr.\n";
    }
    cout << endl;
}

void printTree( const TreeNode* pRoot ) {
    printTreeNode( pRoot );

    if ( pRoot != nullptr ) {
        if ( pRoot->left != nullptr ) printTree( pRoot->left );
        if ( pRoot->right != nullptr ) printTree( pRoot->right );
    }
}

void destroyTree( TreeNode* pRoot ) {
    if ( pRoot != nullptr ) {
        TreeNode* pLeft = pRoot->left;
        TreeNode* pRight = pRoot->right;

        delete pRoot;
        pRoot = nullptr;

        destroyTree( pLeft );
        destroyTree( pRight );
    }
}

// ================ multi-nodes =========================

TreeNodeVec createTreeNodesVec( const ValVec& vals ) {
    TreeNodeVec nodes;
    for ( int i = 0; i < vals.size(); i++ )
        nodes.push_back( createBinaryTreeNode( vals[ i ] ) );
    return nodes;
}

void connectTreeNodesIndex( TreeNodeVec& nodes, int parent_i, int left_i,
                            int right_i ) {
    TreeNode* pLeft = ( left_i != -1 ) ? nodes[ left_i ] : nullptr;
    TreeNode* pRight = ( right_i != -1 ) ? nodes[ right_i ] : nullptr;
    connectTreeNodes( nodes[ parent_i ], pLeft, pRight );
}

void connectTreeNodesIndexVec( TreeNodeVec& nodes, TreeIndexVec& indexes ) {
    int parent_i, left_i, right_i;
    for ( TreeIndex idxs : indexes ) {
        std::tie( parent_i, left_i, right_i ) = idxs;
        connectTreeNodesIndex( nodes, parent_i, left_i, right_i );
    }
}

File: src/Utilities/binary_tree.py
from typing import Tuple

TreeIndex = Tuple[int, int, int]


class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left: TreeNode = None
        self.right: TreeNode = None


# ==== single-node funcs ====


def createBinaryTreeNode(value) -> TreeNode:
    return TreeNode(value)


def connectTreeNodes(parent: TreeNode, left: TreeNode, right: TreeNode) -> None:
    if parent:
        parent.left = left
        parent.right = right


def printTreeNode(node: TreeNode) -> None:
    if node:
        print(f"value of this node is: {node.val}.")

        if node.left:
            print(f"value of its left child is: {node.left.val}.")
        else:
            print("left child is nullptr.")

        if node.right:
            print(f"value of its left child is: {node.right.val}.")
        else:
            print("left child is nullptr.")
    else:
        print("this node is nullptr.")
    print("\n")


def printTree(root: TreeNode) -> None:
    printTreeNode(root)
    if root:
        if root.left:
            printTree(root.left)
        if root.right:
            printTree(root.right)


# ==== multi-node funcs ====


def createTreeNodesVec(vals: list[int]) -> list[TreeNode]:
    nodes = []
    for i in range(len(vals)):
        nodes.append(createBinaryTreeNode(vals[i]))
    return nodes


def connectTreeNodesIndex(
    nodes: list[TreeNode], parent_i: int, left_i: int = -1, right_i: int = -1
) -> None:
    left = nodes[left_i] if left_i != -1 else None
    right = nodes[right_i] if right_i != -1 else None
    connectTreeNodes(nodes[parent_i], left, right)


def connectTreeNodesIndexVec(nodes: list[TreeNode], indexes: list[TreeIndex]):
    for idxs in indexes:
        parent_i, left_i, right_i = idxs
        connectTreeNodesIndex(nodes, parent_i, left_i, right_i)
